internal/cephfs/clone.go

```diff
@@ -0,0 +1,188 @@
+limitations under the License.
+*/
+
+package cephfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	klog "k8s.io/klog/v2"
+)
+
+const (
+	// cephFSCloneFailed indicates that clone is in failed state.
+	cephFSCloneFailed = "failed"
+	// cephFSCloneCompleted indicates that clone is in in-progress state.
+	cephFSCloneInprogress = "in-progress"
+	// cephFSCloneComplete indicates that clone is in complete state.
+	cephFSCloneComplete = "complete"
+)
+
+func createCloneFromSubvolume(ctx context.Context, volID, cloneID volumeID, volOpt, parentvolOpt *volumeOptions, cr *util.Credentials) error {
+	snapshotID := cloneID
+	err := createSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s %v"), snapshotID, err)
+		return err
+	}
+	var (
+		// if protectErr is not nil we will delete the snapshot as the protect fails
+		protectErr error
+		// if cloneErr is not nil we will unprotect the snapshot and delete the snapshot
+		cloneErr error
+	)
+	defer func() {
+		if protectErr != nil {
+			err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			}
+		}
+
+		if cloneErr != nil {
+			if err = purgeVolume(ctx, cloneID, cr, volOpt, true); err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete volume %s: %v"), cloneID, err)
+			}
+			if err = unprotectSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+				// Incase the snap is already unprotected we get ErrSnapProtectionExist error code
+				// in that case we are safe and we could discard this error and we are good to go
+				// ahead with deletion
+				if !errors.Is(err, ErrSnapProtectionExist) {
+					klog.Errorf(util.Log(ctx, "failed to unprotect snapshot %s %v"), snapshotID, err)
+				}
+			}
+			if err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			}
+		}
+	}()
+	protectErr = protectSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+	if protectErr != nil {
+		klog.Errorf(util.Log(ctx, "failed to protect snapshot %s %v"), snapshotID, protectErr)
+		return protectErr
+	}
+
+	cloneErr = cloneSnapshot(ctx, parentvolOpt, cr, volID, snapshotID, cloneID, volOpt)
+	if cloneErr != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone snapshot %s %s to %s %v"), volID, snapshotID, cloneID, cloneErr)
+		return cloneErr
+	}
+	var clone CloneStatus
+	clone, cloneErr = getcloneInfo(ctx, volOpt, cr, cloneID)
+	if cloneErr != nil {
+		return cloneErr
+	}
+	if clone.Status.State == cephFSCloneInprogress {
+		return ErrCloneInProgress{err: fmt.Errorf("clone is in progress for %v", cloneID)}
+	}
+	if clone.Status.State == cephFSCloneFailed {
+		cloneErr = fmt.Errorf("clone %s is in %s state", cloneID, clone.Status.State)
```

Comment:
`cloneErr` gets assigned, but not used/returned except for the `defer`. A comment about that would be good. Needs to return the error too, I think?

Reply:
Sure. logged and also returned.